### PR TITLE
[Core] Wait `ray status` to be ready

### LINKED ISSUE #3129: [Spot] Spot jobs on Azure always says it is preempted before submitting the job
A user reported that a managed spot job on Azure always says it is preempted after the cluster is launched before submitting the job.

This is likely because the `ray cluster` takes longer than other clouds to initalize before `ray status` can return the correct value of the cluster.

### LINKED ISSUE #3134: [Azure] Invalid state string when querying the instance status
A user reported an issue when running `sky serve` with spot replicas, where they got an empty string from the azure cli.
```
E 02-08 21:57:24 replica_managers.py:1000]   File "/home/azureuser/miniconda3/lib/python3.10/site-packages/sky/backends/backend_utils.py", line 1765, in _update_cluster_status_no_lock
E 02-08 21:57:24 replica_managers.py:1000]     node_statuses = _query_cluster_status_via_cloud_api(handle)
E 02-08 21:57:24 replica_managers.py:1000]   File "/home/azureuser/miniconda3/lib/python3.10/site-packages/sky/backends/backend_utils.py", line 1630, in _query_cluster_status_via_cloud_api
E 02-08 21:57:24 replica_managers.py:1000]     node_statuses = cloud.query_status(
E 02-08 21:57:24 replica_managers.py:1000]   File "/home/azureuser/miniconda3/lib/python3.10/site-packages/sky/clouds/azure.py", line 655, in query_status
E 02-08 21:57:24 replica_managers.py:1000]     node_status = status_map[s]
E 02-08 21:57:24 replica_managers.py:1000] KeyError: ''
```

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #3129

This PR also raises proper error for the case when Azure have an invalid state string. Fixes #3134

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
  - [x] reproducible code in #3129
  - [x] `time sky launch -c test-aws --cloud aws -y` (time for waiting ray cluster to start: 6s, this PR: 11s (wait for 5 seconds))
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/backends/backend_utils.py [hunk 1]
```diff
@@ -882,6 +882,10 @@ def write_cluster_config(
                 'dump_port_command': dump_port_command,
                 # Ray version.
                 'ray_version': constants.SKY_REMOTE_RAY_VERSION,
+                # Command for waiting ray cluster to be ready on head.
+                'ray_head_wait_initalized_command':
+                    instance_setup.RAY_HEAD_WAIT_INITALIZED_COMMAND,
```
Comment (cblmemo) [inline sky/backends/backend_utils.py:887]:
```suggestion
                'ray_head_wait_initialized_command':
                    instance_setup.RAY_HEAD_WAIT_INITIALIZED_COMMAND,
```
typo?

File: sky/clouds/azure.py [hunk 2]
```diff
@@ -648,6 +648,9 @@ def query_status(cls, name: str, tag_filters: Dict[str, str],
             original_statuses_list = [original_statuses_list]
         statuses = []
         for s in original_statuses_list:
+            if s not in status_map:
+                raise exceptions.ClusterStatusFetchingError(
```
Comment (cblmemo) [inline sky/clouds/azure.py:652]: nit: `with ux_utils.print_exception_no_traceback():`

File: sky/templates/azure-ray.yml.j2 [hunk 3]
```diff
@@ -169,9 +169,7 @@ setup_commands:
 #   current num items (num SSH connections): 2
 head_start_ray_commands:
   # NOTE: --disable-usage-stats in `ray start` saves 10 seconds of idle wait.
-  - ray stop; RAY_SCHEDULER_EVENTS=0 RAY_DEDUP_LOGS=0 ray start --disable-usage-stats --head --port={{ray_port}} --dashboard-port={{ray_dashboard_port}} --object-manager-port=8076 --autoscaling-config=~/ray_bootstrap_config.yaml {{"--resources='%s'" % custom_resources if custom_resources}} --temp-dir {{ray_temp_dir}} || exit 1;
```
Comment (cblmemo) [inline sky/templates/azure-ray.yml.j2:172]: Just curious: is there any reason to squash the original three lines into one?

File: sky/templates/azure-ray.yml.j2
Comment (cblmemo) [inline sky/templates/azure-ray.yml.j2:1]: Why not add to other clouds like k8s, aws, gcp, vsphere and runpod?

Comment (Michaelvll) [reply sky/templates/azure-ray.yml.j2:1]: Those clouds have adopted the new provisioner API, and has been fixed in the `instance_setup.py`

File: sky/backends/backend_utils.py (hunk 1 above)
Comment (Michaelvll) [reply sky/backends/backend_utils.py:887]: Good catch! Replaced all the occurance. Thanks!

File: sky/templates/azure-ray.yml.j2 (hunk 3 above)
Comment (Michaelvll) [reply sky/templates/azure-ray.yml.j2:172]: Good catch! Changed it back.

**Review by cblmemo — APPROVED** [review]
Thanks! LGTM.

**Pull request merged**